specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1082 @@
+    CheckMsgsRegistry cid.Cid // HAMT[cid]CrossMsgs
+    // Latest nonce of a cross message sent from subnet.
+    Nonce             uint64  
+    // BottomUpNonce of bottomup messages for msgMeta received from che    kpoints.
+    // This nonce is used to mark with a nonce the metadata about cross-net
+    // messages received in checkpoints. This is used to order the
+    // bottom-up cross-messages received through checkpoints.
+    BottomUpNonce     uint64
+    // Queue of bottom-up cross-net messages to be applied.
+    BottomUpMsgsMeta  cid.Cid // AMT[schema.CrossMsgs]
+    // AppliedNonces keep track of the next nonce of the message to be applied.
+    // This prevents potential replay attacks.
+    AppliedBottomUpNonce uint64
+    AppliedTopDownNonce  uint64
+    // Registry with all active atomic executions being orchestrated
+    // by the current subnet.
+    AtomicExecRegistry cid.Cid // HAMT[cid]AtomicExec
+}
+```
+
+> Struct with the information the SCA keeps for each child subnet
+```go
+// Subnet struct kept by the SCA with the information of
+// all of its children.
+type Subnet struct {
+    // ID of the Subnet
+    ID          address.SubnetID 
+    // Parent ID
+    ParentID    address.SubnetID
+    //  Collateral staked for this subnet.
+    Collateral       abi.TokenAmount
+    // List of cross top-down messages committed for the subnet..
+    TopDownMsgs cid.Cid // AMT[ltypes.Messages] 
+    // Latest nonce of cross message submitted to subnet.
+    Nonce      uint64
+    // Amount of native tokens injected in the subnet and
+    // that can be used freely in the subnet.
+    CircSupply abi.TokenAmount 
+    // Status of the checkpoint (`Active`, `Inactive`, etc.)
+    Status     Status
+    // Latest checkpoint committed for the subnet.
+    // Kept for verification purposes.
+    PrevCheckpoint schema.Checkpoint
+}
+```
+
+## Consensus Interface
+Each subnet can run its own implementation of a consensus algorithm. In order for different consensus implementations to operate semlessly with the Filecoin stack, we decouple the core methods of the current consensus in its own interace. In order for a consensus implementation to be usuable in a subnet it needs to implement this interface.
+> Consensus Interface
+```go
+type Consensus interface {
+    // Performs a complete check of a proposed block to either
+    // accept it or reject it.
+    ValidateBlock(ctx context.Context, b *types.FullBlock) (err error)
+    // Light check performed when a block is received through 
+    // the pubsub channel.
+    ValidateBlockPubsub(ctx context.Context, self bool, msg *pubsub.Message) (pubsub.ValidationResult, string)
+
+    // Used by mining processes to assemble and propose a new
+    // signed block.
+    CreateBlock(ctx context.Context, w api.Wallet, bt *api.BlockTemplate) (*types.FullBlock, error)
+    // Return the type of consensus. 
+    Type() hierarchical.ConsensusType
```

Comment:
I would try to remove this from the interface. The consensus should be described by its methods, it should not have to be known by an ID for the host system, ideally, or if it does, we should elaborate on what those are and why this is important. 

As I understand the `SubnetActor` should be the only one that understands how to validate a `Checkpoint` that comes from this consensus. If the correct subnet actor has been registered for a subnet, it should not need a type. 

Now, perhaps the reference implementation of the actor does need a type, because it can handle multiple ones, but in that case the IDs it an handle should be specific to the reference subnet actor, just an example, and not be part of the common `hierarchical` namespace. 

I think this is how it works in Rust, although not in Go. I'm just saying so the spec isn't misleading people into thinking this is a common type they can't extend without buy-in from the whole community.

Reply:
I think I understand your point (you shouldn't need a IANA-style central registry for protocol IDs) but not necessarily your solution. Since the protocols tie into a local (i.e. ex-actor) implementation, how do nodes match the right subnet protocol? Maybe use some UUID-like probabilistically unique ID?

Reply:
I thought that we run:
1. A different node software that implements SuperDuperConsensus in the child subnets. It only communicates with the parent subnet nodes via libp2p gossipsub topics. Nobody knows what SuperDuperConsensus is.
2. Except that the validators deploy an SA actor code into the parent subnet. This actor conforms to the expected SA interface in terms of message handling, but it can also receive SuperDuperCheckpoints and knows how to verify them and send them on to the SCA of the parent subnet.

Reply:
You are totally right, actually we used a similar approach to 1. to integrate Tendermint in the MVP. That being said, we want to ship the node with a built-in catalogue of basic consensus implementations (to lower the barriers for users to join/spawn their subnets), but we should definitely support the approach you suggest.